text/0000-orderer-snapshot.md

```diff
@@ -0,0 +1,316 @@
+
+This RFC introduces a notion of a **ledger snapshot for OSNs** (ordering service nodes). With this notion two new logical operations appear at the orderer REST API and in the CLI tool *osnadmin*:
+- Create a snapshot,
+- Join channel from a snapshot.
+
+Ledger snapshot is a set of files containing all the necessary information for OSN to join a channel from an arbitrary block in the chain without loss of any functionality except the ability to deliver an earlier block.
+
+A **local orderer administrator** is a role that was introduced in RFC-0004 "Channel participation API without system channel" and is currently implemented with the help of "Admin Configuration" section in the OSN YAML config file (<Config.Admin.TLS>). Only the identity associated with this role is able to invoke the API defined in this RFC, that is existing **channel participation API** and newly defined **ledger management API**.
+
+## Guide: Create snapshot
+
+To create a ledger snapshot use the following REST API (optional block_number should be supplied as a parameter in the request body):
+
+    POST /ledger/v1/{channelID}/snapshot
+
+ or the following CLI command:
+
+    $ osnadmin ledger snapshot -c <channelID> [-b <block_number>]
+
+If the optional parameter <block_number> is omitted, snapshot will be created from the latest block, i.e. from the block number equal to ledgerHeight-1.
+
+A process of snapshot creation will start immediately if the block number is omitted, or if the block number is specified and is less than current ledger height. If the block number is specified and is in future (>= ledger height) then the operation will immediately fail with a reasonable error message.
+
+The process of snapshot creation is a synchronous one; upon the successful completion of the "ledger snapshot" API call the snapshot itself is ready and can be accessed in the following directory in the file system:
+```
+<Config.FileLedger.snapshots.rootDir>/completed/<channelID>/<block_number>
+```
+
+## Guide: Join channel from a snapshot
+
+To join a channel from a snapshot use the following REST API (files from the snapshot must be placed in the body, using multipart/form-data content-type):
+
+    POST /participation/v1/joinbysnapshot
+
+ or the following CLI command:
+
+    $ osnadmin channel joinbysnapshot -s <snapshot_dir>
+
+There is no need to specify {channelID} when using the above API, as it will be extracted from the snapshot content.
+
+To check the status of the "join by snapshot" operation use the following REST API:
+
+    GET /participation/v1/channels/{channelID}
+
+ or the following CLI command:
+
+    $ osnadmin channel list -c <channelID>
+
+## Guide: Prune block ledger of an existing channel
+[guide-prune-block-ledger-of-an-existing-channel]: #guide-prune-block-ledger-of-an-existing-channel
+
+To prune an existing channel ledger do the following:
+ - create a new snapshot at the desired block number,
+ - remove the OSN from the channel,
```

Comment:
I don't understand why we're coupling here a channel scoped operation and a local operation.
If we want to prune an OSN then why not just prune it? 

Removing the OSN from the channel requires config update and joining it "from a snapshot" doesn't actually "bring it back" to the channel. 

I think we should include the prune step that is discussed in the drawback section in this RFC and not in the next one.

Reply:
In a previous discussion with @manish-sethi it was decided that it is better to limit the scope of this RFC to snapshots only. On-the-fly prune operation is not a trivial one (mostly because of the concurrency issues), and snapshot functionality does not really depend on it.
https://github.com/hyperledger/fabric-rfcs/pull/44#discussion_r603650437

> Removing the OSN from the channel requires config update and joining it "from a snapshot" doesn't actually "bring it back" to the channel.

Snapshot functionality works only when the system channel does not exist, so if an admin is following these steps when system channel is present in the setup, the very first step (snapshot creation) will fail immediately.
(note: probably I should have mentioned this explicitly, that the described sequence of commands works only when the system channel is not present)

Also by removing OSN from channel I mean using the channel participation API "channel remove". This API does not produce new config block, it simply deletes the local ledger for the channel on the local filesystem of the OSN.
(note: again, probably I should have mentioned this explicitly to avoid confusion)

So if we use newly defined "channel joinbysnapshot" API in the last step, the OSN will pick up the last known config block which has this orderer listed as one of the consenters.

-------

In general this section of the *Guide-level explanation* is there only to demonstrate how this new snapshot feature can be used to prune the block ledger on the disk. This section does not mandate anything, it is just a usage example.

Reply:
Oh I see I thought you wanted to do a config update that removes the OSN from the channel.
What you really meant was - _remove the channel from the OSN_, not _remove the OSN from the channel_. 

However let me ask you something - what if that node is a leader? What do we do then? It may destabilize the cluster. Perhaps we should try to [transfer the leadership](https://github.com/etcd-io/etcd/blob/main/raft/node.go#L180) in case that node is a leader for the channel?

Reply:
> What you really meant was - remove the channel from the OSN, not remove the OSN from the channel.

Yes, you are right.
So I improved the wording here to avoid confusion and re-pushed.

> However let me ask you something - what if that node is a leader? What do we do then? It may destabilize the cluster. Perhaps we should try to transfer the leadership in case that node is a leader for the channel?

Unfortunately an admin does not have an API to transfer the leadership. So they will have to wait a bit for a new re-election to occur by timeout.

All in all, this is one more argument in favour of the future "pruning on-the-fly" RFC.

Reply:
I was trying to say that the node itself may transfer the leadership if it is the leader and only then proceed with deleting the channel from itself.

Reply:
Ah, I see.
Yes, I agree, it would be a nice improvement to the implementation of RemoveChannel() API.

I personally think that this is not in the scope of the RFC, but a separate JIRA ticket makes sense.